### ISSUE TITLE
Initialization of static arrays in storage

### ISSUE THREAD

**Issue by cicr99** [NethermindEth/warp#863]
The following contract:
```
// SPDX-License-Identifier: AGPL-3.0-only
pragma solidity ^0.8.14;

contract WARP {
  uint256[100000000] d;
}
```
fails with error: `FATAL ERROR: Ineffective mark-compacts near heap limit Allocation failed - JavaScript heap out of memory`
This only happens when the size of the array is big enough, for smaller size values it works fine.

The issue probably has to do with line https://github.com/NethermindEth/warp/blob/develop/src/passes/storageAllocator.ts#L60 in `StorageAllocator` pass, since it tries to create a new typeNode for each element of the array

**Comment by LuisLara-UH** [comment]
This issue lies on line https://github.com/NethermindEth/warp/blob/develop/src/utils/defaultValueNodes.ts#L147 in `VariableDeclarationInitializer` pass. When it substitutes the static array for a tuple, it creates an ExpressionList containing an expression for each element of the array, which overflows the Javascript heap.

I think the solution for this could be to increase the maximum size of Node heap. But even with this, with larger arrays it is going to overflow the heap too. So, the solution also should have some throwing error for larger arrays.

@cicr99 could you please assign this task to me?

**Comment by cicr99** [comment]
The issue you are talking about is different from this one and it's being worked on already in [#821](https://github.com/NethermindEth/warp/issues/821). That one refers to the initialization of static arrays in memory and this one has to do with storage.

The thing with this contract is that `d` is a state variable so we handle it first on the `StorageAllocator` pass. I looked into the bug a bit further and the transpilation breaks at this [line](https://github.com/NethermindEth/warp/blob/242439ab73029cee4de63ac3898f8bee0cdb4fff/src/utils/cairoTypeSystem.ts#L74):
```
return new CairoTuple(Array(narrowedLength).fill(elementType));
```
This is called from the `StorageAllocator` pass in the line I had mentioned in the previous comment. Maybe the solution for this is to create a different type of node rather than using `CairoTuple`. I think @rodrigo-pino had an idea about it, so maybe he can comment some more on the matter

**Comment by rodrodros** [comment]
The problem is that we treat solidity static arrays as cairo tuples. But static arrays are a subset of tuples, because though they can have `n` elements, each element must be of a single type `T`. The solution to this bug would be taking advantage of this characteristic of static arrays, where there is no need to initialize the whole `n` elements with type `T`.

Several ways I think of approaching this issue (I haven't investigated them fully):
1. Optimzing the `CairoTuple` type to group similar types, if it has 't' different types, then  it has inside a list of size `t` but to the outside user it still a tuple of `n` elements with `t <= n`.
2. Creating a type which is related (as, an ancestor or a successor) of a `CairoTuple` and has only type and the size. (It's important for it to be related to avoid compatibility issues further down the line)
3. Same idea as the previous one, but the types has no relationship between them. 

I believe (2) and (3) would result in a much harder task. I would opt for (1) for the moment being since it does not exclude any other possible solution to be applied later, fix the bug right away and I think it's a cool optimization for tuple in generals.

@LuisLara-UH  @cicr99

**Comment by cicr99** [comment]
I would go with option 3, even though it looks like these two concepts are related, they actually wouldn't share any of the implementation of their members. Going with option 1 is something I don't like very much because you'll need to handle the fact that same types might not be grouped together, which could be solved using a range system or something like that, but it sounds to me like a messier solution.
However, I reviewed the code and apparently `CairoTuple` class is only being used for static arrays, it was just implemented that way for it to be as generic as possible. So another idea could be to replace the implementation of this class to handle only static arrays.

**Comment by LuisLara-UH** [comment]
Yes, right now `CairoTuple` in the code is not being used since I am using `CairoStaticArray` for static arrays. I thought it could be useful to let the implementation of `CairoTuple` because we could need that generic implementation later(for tuples maybe?)

**Comment by cicr99** [comment]
Closed by #873
